lib/ruby/lib/alces/stack/status/run.rb

```diff
@@ -19,138 +19,116 @@
+        lambda_proc = lambda { |lambda_hash| lambda_hash[:nodename] }
+        a = Alces::Stack::Iterator.run(group, lambda_proc, nodename: nodename)
+        a.is_a?(Array) ? a : [a]
+      end
+    end
+
     module Status
       class Run
-        include Alces::Tools::Execution
-
         def initialize(options={})
-          @opt = Options.new(options)
+          @opt = RunOptions.new(options)
+          raise "Can not specify both -g and -n" if @opt.nodename? == @opt.group?
         end
 
-        class Options
-          def initialize(options)
-            @options = options
-            assert_preconditions
-          end
-
-          def assert_preconditions
-            raise InputError.new "Requires: -n xor -g" unless group? ^ nodename?
-          end
-          class InputError < StandardError; end
-
-          def nodes
-            lambda_proc = lambda { |options| options[:nodename] }
-            @nodes ||= lambda {
-                a = Alces::Stack::Iterator
-                      .run(group, lambda_proc, nodename: nodename)
-                a = [a] unless a.is_a? Array 
-                a
-              }.call
-          end
-
-          def cmds
-            [:power, :ping]
-          end
-
-          def method_missing(s, *a, &b)
-            if @options.key?(s)
-              @options[s]
-            elsif s[-1] == "?"
-              !!@options[s[0...-1].to_sym]
-            else
-              super
-            end
-          end
+        def run!
+          start_monitor
+          collect_data
+          @monitor.thread.join
         end
-
-        def set_signal
-          Signal.trap("INT") {
-            if @int_once
-              @monitor.kill
-            else
-              @int_once = true
-              begin
-                @monitor.wait
-              rescue
-              end
-            end
-            File.delete(@report_file) if File.exist?(@report_file.to_s)
-            Kernel.exit
-          }
+        
+        def start_monitor
+          @monitor = Alces::Stack::Status::Monitor.new({
+              nodes: @opt.nodes,
+              cmds: @opt.cmds,
+              thread_limit: @opt.thread_limit,
+              time_limit: @opt.time_limit
+            })
+          @monitor.start
         end
 
-        def set_logging
-          status_log = Alces::Stack::Log.create_log("/var/log/metalware/status.log")
-          status_log.progname = "status"
-          Alces::Stack::Status::Monitor.log = status_log
-          Alces::Stack::Status::Task.log = status_log
+        def collect_data
+          data = {}
+          empty_count = 0
+          while data["FINISHED"] != true
+            data = get_finished_data
+            if data.empty?
+              empty_count += 1
+              raise DataIncomplete if empty_count > 100 && @monitor.thread.stop?
+            elsif data["FINISHED"] != true
+              display_data data
+              empty_count = 0
+            end
+            sleep 0.1
```

Comment:
Why the magic sleep? Could be worth adding a comment to this with the reason as the purpose isn't obvious, and similarly for other magic sleeps.